### ISSUE TITLE
Add automated labelling to staged-recipes

### ISSUE THREAD

**Issue by BastianZim** [conda-forge/staged-recipes#18023]
This is a proposal to add automated labelling to the staged-recipes repo using GitHub actions.
I'm opening this as an issue instead of a PR as I wanted to get some input first what everybody else thinks. 
As conda-forge is generally very hands-on I didn't want to add too much automation and thought that the following would be a good start:

1. Label PRs based on the requested team. 
  This is pretty straightforward and would only add a label with the pinged team. It would make browsing PRs easier and quicker as you can see if it's in your domain or not.
  Additionally, even PRs by new contributors that cannot yet ping a team are labelled and can be marked as ready for review.
  
    The GitHub actions yaml would be something like this:
    ```yaml
    name: Label PRs

    on:
      issue_comment:
        types: [created]

    jobs:
      label-pr:
        name: Label PR
        if: ${{ github.event.issue.pull_request }}
        runs-on: ubuntu-latest
        strategy:
          matrix:
           include:
             - label: 'staged-recipes'
               comment: '@conda-forge/staged-recipes'

             - label: 'python'
               comment: '@conda-forge/help-python'

             - label: 'python-c'
               comment: '@conda-forge/help-python-c'

             - label: 'r'
               comment: '@conda-forge/help-r'

             - label: 'java'
               comment: '@conda-forge/help-java'

             - label: 'nodejs'
               comment: '@conda-forge/help-nodejs'

             - label: 'c-cpp'
               comment: '@conda-forge/help-c-cpp'

             - label: 'perl'
               comment: '@conda-forge/help-perl'

             - label: 'julia'
               comment: '@conda-forge/help-julia'

             - label: 'ruby'
               comment: '@conda-forge/help-ruby'

        steps:
          - name: team-label
            if: contains(github.event.comment.body, matrix.comment)
            env:
              label: ${{ matrix.label }}
            uses: actions/github-script@v6
            with:
              script: |
                github.rest.issues.addLabels({
                  issue_number: context.issue.number,
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  labels: [process.env.label, 'review-requested']
                })
    ```

    I tested the code locally and noticed a couple of small issues:
    1. conda-forge/help-python-c also includes conda-forge/help-python as it checks if the string is included.
    2. Currently runs for every label as the `if` is only checked during the run. Maybe there is a better way.

2. Add review-requested lables
 This would be set with the above action and the reviewer can then remove it again manually (can also be automated probably).
 
    The benefit of this is that sometimes PRs are missed because no-one reviewed them which would now be visible. So instead of going to each PR and checking if it has been reviewed yet, it can be seen from the overview page.
  
    Additionally, the staged-recipes team can see if the reason another team member commented was because they reviewed the PR or are the author and actually still need a review.

**Comment by BastianZim** [comment]
CC @conda-forge/core

**Comment by beckermr** [comment]
Lgtm!

**Comment by jaimergp** [comment]
> conda-forge/help-python-c also includes conda-forge/help-python as it checks if the string is included.

You can add a second clause `and !contains(matrix.comment + '-')` or similar?

Besides that, I think having that kind of parallelism is overkill. I'd say it be better API-usage wise to roll a for-loop in the GH script step. Launching 10 instances per comment on every open PR is a bit overkill imo. 

I love this idea but we should think how to implement it in a way that doesn't affect the API usage limits too much!

**Comment by BastianZim** [comment]
> You can add a second clause and !contains(matrix.comment + '-') or similar?

Good idea!

> I love this idea but we should think how to implement it in a way that doesn't affect the API usage limits too much!

Definitely. This is more a proof of concept so far. 😄  Implementation should be more optimised.

**Comment by beckermr** [comment]
I wouldn't worry too much about api limits. If we can encounter an issue we can fix it.

**Comment by BastianZim** [comment]
@jaimergp I'm not sure how `and !contains(matrix.comment + '-')` can work because it could be that a comment includes more than the ping command. So `-` could be added but not be a part of the ping command.

**Comment by BastianZim** [comment]
One other question: The `review-requested` would need to be manually removed. Do we need a vote on that since it involves actively changing something?

We could also automate that using the following logic:

```
If commenter in [staged-recipes-team] but not in [author]:
  Remove label
```

Which one do we like more?

**Comment by xhochy** [comment]
> We could also automate that using the following logic:
> 
> ```
> If commenter in [staged-recipes-team] but not in [author]:
>   Remove label
> ```

I really like this automation and would also like to see the opposite, for the label `Awaiting author contribution` (remove on comment-from-author or commit).

**Comment by jaimergp** [comment]
> @jaimergp I'm not sure how `and !contains(matrix.comment + '-')` can work because it could be that a comment includes more than the ping command. So `-` could be added but not be a part of the ping command.

I think I meant to say this, sorry!

```diff
- if: contains(github.event.comment.body, matrix.comment)
+ if: contains(github.event.comment.body, matrix.comment) and !contains(github.event.comment.body, matrix.comment + '-')
```

But maybe I am not seeing something super obvious, so in that case this regard this suggestion!

**Comment by BastianZim** [comment]
One other thing: We could print a Grayskull output the first time the recipe is ready for review, as that's most often the first thing, at least I, use for reviewing. What's everybody's opinion on that?

**Comment by BastianZim** [comment]
I have created an initial PR here: https://github.com/conda-forge/staged-recipes/pull/18030

Any feedback welcome!

**Comment by BastianZim** [comment]
Hi everybody,

unfortunately I had to open the issue again because we ran into some trouble and wanted to get some input on how to go forward.

Initially, the workflow would remove the review-requested label when someone interacted with the PR and was part of the staged-recipes team. 
For that, a personal access token (PAT) is needed which cannot be supplied to a fork so that version did not work.

@carterbox then worked on an alternative, where the pinged team gets assigned to the PR and, after a review, the workflow triggers based on the fact that no team is requested anymore. Unfortunately, teams cannot be assigned to a PR without a PAT: https://github.community/t/bug-github-action-cannot-assign-teams/131186

So, we are now in a situation where we are unable to proceed with a check for membership because any way that we can come up with requires a PAT which is not available on PRs from forks.

Therefore, the only solution we can currently come up with is to remove the check and just remove the review-requested label whenever someone comments, no matter whether they are a member of staged-recipes or not. Otherwise we can also just not remove the label at all but that requires more work on the reviewer side as they need to remove it manually.

The only other way would be to write a [bot](https://github.com/probot) but that needs to be hosted on a server somewhere and I'm not sure if that resource is available. 

What's everybody's opinion on this?

**Comment by beckermr** [comment]
We could not automate the label removing step and just the reviewer do it by hand?

**Comment by BastianZim** [comment]
> We could not automate the label removing step and just the reviewer do it by hand?

Sure, we just weren't sure if everybody wants to do that?

**Comment by beckermr** [comment]
That seems fine to me. Let's not the perfect be the enemy of the good or w/e.

**Comment by carterbox** [comment]
> We could not automate the label removing step and just the reviewer do it by hand?

@BastianZim I was unable to use the `workflow_run` trigger, so as Matthew suggested, I worked around the fork permissions limitations by using different triggers from our original plan. Instead of using `pull_request_review: submitted` which runs without permission on the fork, I switched to `pull_request_target: unlabeled`, so the reviewer has to manually remove the 'review-requested' label.

I tested this approach on our test PR and let it run live for an hour today. The only thing I haven't tested is "what happens if someone outside of staged-recipes tries to remove the review-requested label?" I think it's working, but I'm going to disable the action for now. When you get a chance, look at the code and let it run for a few hours while monitoring on Monday?

Initial thoughts are that the action runs for EVERY comment. Maybe we should skip if the comments are from other bots? I don't know if bots should be allowed to request review of a PR.
